pilz_control/test/unittest_cartesian_speed_monitor.cpp

```diff
@@ -0,0 +1,255 @@
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, 0.0, SPEED_COMPARISON_TOLERANCE);
+
+  link = model_->getLinkModel("link3");
+  double expected_velocity = angular_displacement / time_delta;
+  speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, expected_velocity, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedLinearInTime)
+{
+  robot_state::RobotStatePtr state(new robot_state::RobotState(model_));
+  state->setToDefaultValues();
+  state->updateLinkTransforms();
+
+  robot_state::RobotStatePtr state2(new robot_state::RobotState(model_));
+  state2->setToDefaultValues();
+  std::vector<double> angular_displacements{ 0.1, 0.1 };
+  state2->setVariablePositions(angular_displacements);
+  state2->updateLinkTransforms();
+
+  moveit::core::LinkModel* link = model_->getLinkModel("link3");
+  double time_delta = 0.1;
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+
+  double factor = 2.0;
+  time_delta *= factor;
+  double speed2 = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+
+  EXPECT_NEAR(speed2 * factor, speed, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testBelowLimitMoveJoint2)
+{
+  CartesianSpeedMonitor monitor(joint_names_, model_);
+  monitor.init();
+
+  double angular_displacement = 0.1;
+  double time_delta = 0.2;
+  double limit = angular_displacement / time_delta + SMALL_SKIP;
+  EXPECT_TRUE(monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.0 }, { 0.0, angular_displacement }, time_delta, limit));
+
+  limit = angular_displacement / time_delta - SMALL_SKIP;
+  EXPECT_FALSE(monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.0 }, { 0.0, angular_displacement }, time_delta, limit));
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testBelowLimitMoveJoint1)
+{
+  CartesianSpeedMonitor monitor(joint_names_, model_);
+  monitor.init();
+
+  double angular_displacement = 0.1;
+  double time_delta = 0.2;
+  double limit = angular_displacement / time_delta + SMALL_SKIP;
+  EXPECT_TRUE(monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.0 }, { angular_displacement, 0.0 }, time_delta, limit));
+
+  limit = angular_displacement / time_delta - SMALL_SKIP;
+  EXPECT_FALSE(monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.0 }, { angular_displacement, 0.0 }, time_delta, limit));
+
+  // move link2/3 out of perpendicular position
+  limit = angular_displacement / time_delta + SMALL_SKIP;
+  EXPECT_FALSE(
+      monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.5 * M_PI }, { angular_displacement, 0.5 * M_PI }, time_delta, limit));
+}
```

Comment:
I reduced the code duplication by using a for-loop.

I would keep the model. It intentionally has a chain with two continuous joints. When moving a single joint, we expect different outcome for different joints. Especially, we can cover two cases: 2 Links are fixed in perpendicular position, or not.
This would not be possible with the new model.